authority/provisioner/options.go

```
@@ -0,0 +1,136 @@
+
+	// TemplateData is a JSON object with variables that can be used in custom
+	// templates.
+	TemplateData json.RawMessage `json:"templateData"`
+}
+
+// HasTemplate returns true if a template is defined in the provisioner options.
+func (o *X509Options) HasTemplate() bool {
+	return o != nil && (o.Template != "" || o.TemplateFile != "")
+}
+
+// TemplateOptions generates a CertificateOptions with the template and data
+// defined in the ProvisionerOptions, the provisioner generated data, and the
+// user data provided in the request. If no template has been provided,
+// x509util.DefaultLeafTemplate will be used.
+func TemplateOptions(o *Options, data x509util.TemplateData) (CertificateOptions, error) {
+	return CustomTemplateOptions(o, data, x509util.DefaultLeafTemplate)
+}
+
+// CustomTemplateOptions generates a CertificateOptions with the template, data
+// defined in the ProvisionerOptions, the provisioner generated data and the
+// user data provided in the request. If no template has been provided in the
+// ProvisionerOptions, the given template will be used.
+func CustomTemplateOptions(o *Options, data x509util.TemplateData, defaultTemplate string) (CertificateOptions, error) {
+	opts := o.GetX509Options()
+	if data == nil {
+		data = x509util.NewTemplateData()
+	}
+
+	if opts != nil {
+		// Add template data if any.
+		if len(opts.TemplateData) > 0 {
+			if err := json.Unmarshal(opts.TemplateData, &data); err != nil {
+				return nil, errors.Wrap(err, "error unmarshaling template data")
+			}
+		}
+	}
+
+	return certificateOptionsFunc(func(so SignOptions) []x509util.Option {
+		// We're not provided user data without custom templates.
+		if !opts.HasTemplate() {
+			return []x509util.Option{
+				x509util.WithTemplate(defaultTemplate, data),
+			}
+		}
+
+		// Add user provided data.
+		if len(so.TemplateData) > 0 {
+			userObject := make(map[string]interface{})
+			if err := json.Unmarshal(so.TemplateData, &userObject); err != nil {
```

Should we at least print something here? Warning? otherwise it's just a silent failure.